### PR TITLE
feat: break render islands single task into to multiples

### PR DESCRIPTION
**Pull request by igorbrasileiro**
# Pull Request Context
The fresh hydration runs inside a single tasks, that means all plugins and all islands js execution will run in one tasks and probably will create a long task. Long Task, following the google definition, is every task the takes more than 50ms. This behavior impacts the total blocking time (TBT) and the Time to Interactive (TTI) metrics from Lighthouse (LH)

This algorithm increase linearly with the number of plugins/islands

## What this Pull Request does?
Knowing the problem above, this pull request suggest breaking the revive function task, that render all islands, into multiple tasks, one by island using the scheduler.postTask api or to setTimeout zero as fallback. Here a reference about "breaking long tasks" https://web.dev/optimize-long-tasks/
 What's the benefit of this:
1. for each island the browser creates a task, so, each island has 50ms additional time, this impact directly the TBT and TTI;
2. Also, browsers that implements the scheduler, can prioritize user interaction over hydration/revive, and this gives the site user a better user experience;

> ⚠️ This changes could run under an experimental flag, but I don't know what's the best approach to do this. I left a TODO warning about it.

## Results
I ran 30 PageSpeed Insights test via API, using https://github.com/igorbrasileiro/psi-sample-test, for each deploy preview with strategy mobile.

Example 1:
Before the change: https://deco-sites-fashion-xpv2mkxz84gg.deno.dev/
| Metric | Mean | Standard deviation | Confidence Interval (95%) |
|--------|--------|--------|--------|
| Cumulative Layout shift (CLS) | 0.00 | 0.00 | [0.00, 0.00] |
| First Contentful Paint (FCP) | 1836.81 | 1.43 | [1836.30, 1837.32] |
| Largest Contentful Paint (LCP) | 2201.81 | 39778.79 | [-12032.84, 16436.47] |
| Time to Interactive (TTI) | 2123.17 | 14203.80 | [-2959.59, 7205.94] |
| Total Blocking Time (TBT) | 184.18 | 6179.87 | [-2027.26, 2395.62] |
| Performance score | 0.916 | 0.002511 | [0.914768, 0.916565] |
| JavaScript Execution Time | 719.95 | 23649.72 | [-7742.99, 9182.90] |
| Speed Index | 4085.94 | 5382052.26 | [-1921856.72, 1930028.61] |

After the change: https://deco-sites-fashion-pmxene9e23eg.deno.dev/
| Metric | Mean | Standard deviation | Confidence Interval (95%) |
|--------|--------|--------|--------|
| Cumulative Layout shift (CLS) | 0.00 | 0.00 | [0.00, 0.00] |
| First Contentful Paint (FCP) | 1836.14 | 0.00 | [1836.14, 1836.14] |
| Largest Contentful Paint (LCP) | 2136.14 | 0.00 | [2136.14, 2136.14] |
| Time to Interactive (TTI) | 2087.71 | 27.62 | [2077.83, 2097.59] |
| Total Blocking Time (TBT) | 108.10 | 164.64 | [49.18, 167.02] |
| Performance score | 0.963 | 0.000021 | [0.962992, 0.963008] |
| JavaScript Execution Time | 665.85 | 566.01 | [463.30, 868.39] |
| Speed Index | 2656.46 | 238907.39 | [-82835.45, 88148.37] |

As you can see, the TTI and TBT the mean is lower and the std deviation too.

## References
https://web.dev/optimize-long-tasks/
https://web.dev/tbt/
https://pagespeed.web.dev/
https://github.com/igorbrasileiro/psi-sample-test

### CODE REVIEW

**Comment by igorbrasileiro** [comment]
The project and deploy preview used as example comes from https://github.com/deco-sites/fashion/pull/46

**Comment by igorbrasileiro** [comment]
The flamegraph differences
before the change: you can see one task for the plugins + revive + render islands
image

after the change: you can see one task for plugins + revive and other small tasks for render the islands 
image

**Comment by guifromrio** [comment]
@lucacasonato @ry sorry to bother — did you get a chance to look at this one? We understand it's a low hanging fruit which could help our PageSpeed performance a bit.

**Comment by tlgimenes** [comment]
Hey folks, any news on this one? This seems to be a free lunch in terms of lighthouse scores.

File: src/runtime/main.ts [hunk 1]
```diff
@@ -41,14 +44,20 @@ export function revive(islands: Record<string, ComponentType>, props: any[]) {
...
+        render(
+          h(islands[id], props[Number(n)]),
+          createRootFragment(
+            parent! as HTMLElement,
+            children,
+            // deno-lint-ignore no-explicit-any
+          ) as any as HTMLElement,
+        );
+      };
+      // TODO: Run this under a flag
```
Comment (iuioiua) [inline src/runtime/main.ts]:
```suggestion
```
I think this is fine to run without a flag.

File: src/runtime/main.ts [hunk 2]
```diff
@@ -41,14 +44,20 @@ export function revive(islands: Record<string, ComponentType>, props: any[]) {
...
+      const _render = () => {
+        render(
+          h(islands[id], props[Number(n)]),
+          createRootFragment(
+            parent! as HTMLElement,
+            children,
+            // deno-lint-ignore no-explicit-any
+          ) as any as HTMLElement,
+        );
+      };
```
Comment (iuioiua) [inline src/runtime/main.ts]:
```suggestion
      function _render() {
        render(
          h(islands[id], props[Number(n)]),
          createRootFragment(
            parent! as HTMLElement,
            children,
            // deno-lint-ignore no-explicit-any
          ) as any as HTMLElement,
        );
      }
```
Nit: arrow functions are only really used with anonymous functions in this codebase.

**Review by iuioiua — CHANGES_REQUESTED** [review]

File: src/runtime/main.ts (hunk 2 above)
Comment (iuioiua) [reply src/runtime/main.ts]: Could you please also resolve the merge conflict?

**Comment by iuioiua** [comment]
Note: [`scheduler.postTask()` is well-supported, at 70% of all users](https://caniuse.com/mdn-api_scheduler_posttask). My only concern is the backup use of `setTimeout()`, which could impede performance for a large number of islands. However, the benefit of doing so will likely outweigh any headaches, so +1 from me.

If I'm wrong, alternatively, we can instead have the backup function be the old-fashioned `_render()`.

**Review by iuioiua — APPROVED** [review]
I've made the suggested changes, except for turning `_render()` into a normal function to avoid a lint error. Now, LGTM! Thanks, @igorbrasileiro.

**Review by lucacasonato — APPROVED** [review]
LGTM

**Comment by vibegui** [comment]
Great work @igorbrasileiro, thanks for the kind attention @iuioiua and @lucacasonato !

**Pull request merged**